### PR TITLE
`$not` operator

### PR DESCRIPTION
**Pull request by maheshrajamani**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ArrayComparisonOperator.java
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ArrayComparisonOperator.java:8]: This is used only for negation for $all, not exposed as a filter operator

Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ArrayComparisonOperator.java:8]: Maybe add this exact explanation in Java comment, as that is very helpful.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ComparisonExpression.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ComparisonExpression.java:56]: If this is done to flip between 1 and -1, more efficient way would be to use `BigDecimal.negate()` (https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#negate-- )

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java:81]: Is it possible this could be called for `NOT`? And if so, is "do nothing" what should be done (if it shouldn't be called, should probably throw `IllegalStateException` in `else` case.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/model/impl/matcher/FilterableResolver.java
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/model/impl/matcher/FilterableResolver.java:314]: Ah. So here the earlier negative value for `$size` (multiply by -1) is converted to "not equals $size". Makes sense.

**Review by tatu-at-datastax — APPROVED** [review]
LGTM: added some questions, but nothing blocking merge.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java
Comment (maheshrajamani) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java:81]: We resolve depth first handled `NOT` by flipping the operators and moved the expression to the parent logical expression. Because of that flip() will not have any `NOT` logical expression below it.

Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java:81]: Ok, thanks.

**Review by tatu-at-datastax — APPROVED** [review]
LGTM

**Review by vkarpov15 — COMMENTED** [review]
I'm curious about why `$not` is an array. In MongoDB, $not is an object like the following:

```javascript
find({
  num: {
    $not: {
      $lt: 42
    }
  }
}); // Equivalent to `find({ num: { $gte: 42 } })`
```

Is there a reason for making $not an array?

**Comment by maheshrajamani** [comment]
> I'm curious about why `$not` is an array. In MongoDB, $not is an object like the following:

I made it in line with `$and` and `$or` operator. It does `and` of conditions below it. Do we have to change this?

**Comment by tatu-at-datastax** [comment]
> > I'm curious about why `$not` is an array. In MongoDB, $not is an object like the following:
> 
> I made it in line with `$and` and `$or` operator. It does `and` of conditions below it. Do we have to change this?

Hmmmh. Good point, thank you for pointing it out. It is interesting Mongo has this discrepancy: `$and` and `$or` take Array, `$not` for some reason Object. Go figure.

Earlier we wanted to be highly Mongo-compatible, and I would have suggested we follow its lead.
But now we are moving away from that.
Still, we are mostly extending things and not changing handling of features Mongo has.
So I am not 100% sure which way we should go.

I think we need to discuss this on our daily meeting @maheshrajamani see what Aaron thinks.

**Comment by tatu-at-datastax** [comment]
Based on discussion with @amorton, I think the preferred option is to use JSON Object, same as how Mongo behaves.
I guess it also makes sense logically, come to think of it: `not` is typically unary operation whereas `and` and `or` are binary. So taking in Array would be more likely implicit "not and" (and possibly more confusing).

**Comment by vkarpov15** [comment]
Another reason why we should make `$not` a unary operator is that [Mongoose's filter casting logic assumes `$not` is unary](https://github.com/Automattic/mongoose/blob/265741752a2834db806def4520091b3b318f73f4/lib/cast.js#L331-L350). In the following script, Mongoose throws a CastError `CastError: Cast to number failed for value "[ { '$gt': 42 } ]" (type Array) at path "age" for model "Test"`:

```javascript
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://127.0.0.1:27017/mongoose_test');

  const schema = new mongoose.Schema({ age: Number });
  const Test = mongoose.model('Test', schema);

  await Test.find({ age: { $not: [{ $gt: 42 }] } });
  console.log('Done');
}
```

However, `await Test.find({ age: { $not: { $gt: 42 } } });` (object rather than array for `$not`) casts correctly. So if we want to be compatible with what Mongoose expects, then `$not` should be unary.

**Comment by maheshrajamani** [comment]
@vkarpov15 @tatu-at-datastax Have made the `$not` to accept one object as value. If you want to do another review of it.

**Review by tatu-at-datastax — APPROVED** [review]

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java
Comment (vkarpov15) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java:1856]: Nice, I'm glad you tested this. I checked whether multiple `$not` works in MongoDB and Mongoose: MongoDB supports it, but Mongoose doesn't. Something for Mongoose to fix.

**Pull request merged**
